gc/gc-iceberg/src/main/java/org/projectnessie/gc/iceberg/ExpireContentsProcedure.java

```diff
@@ -0,0 +1,293 @@
+  public static final String OUTPUT_EXPIRED_DATA_FILES_TYPE = "deleted_files_type";
+  public static final String OUTPUT_EXPIRED_DATA_FILES_COUNT = "deleted_files_count";
+  public static final String OUTPUT_EXPIRED_FILES_LIST = "deleted_files_list";
+
+  public static final StructType OUTPUT_TYPE =
+      new StructType(
+          new StructField[] {
+            new StructField(OUTPUT_CONTENT_ID, DataTypes.StringType, true, Metadata.empty()),
+            new StructField(
+                OUTPUT_EXPIRED_DATA_FILES_TYPE, DataTypes.StringType, true, Metadata.empty()),
+            new StructField(
+                OUTPUT_EXPIRED_DATA_FILES_COUNT, DataTypes.IntegerType, true, Metadata.empty()),
+            new StructField(
+                OUTPUT_EXPIRED_FILES_LIST,
+                DataTypes.createArrayType(DataTypes.StringType),
+                true,
+                Metadata.empty())
+          });
+
+  public enum ExpiredContentType {
+    ICEBERG_MANIFEST,
+    ICEBERG_MANIFESTLIST,
+    DATA_FILE
+  }
+
+  public ExpireContentsProcedure(TableCatalog currentCatalog) {
+    super(currentCatalog);
+  }
+
+  @Override
+  public ProcedureParameter[] parameters() {
+    return PARAMETERS;
+  }
+
+  @Override
+  public StructType outputType() {
+    return OUTPUT_TYPE;
+  }
+
+  @Override
+  public String description() {
+    return String.format(
+        "Expires the Iceberg snapshots that are collected by '%s' procedure.",
+        IdentifyExpiredContentsProcedure.PROCEDURE_NAME);
+  }
+
+  @Override
+  public InternalRow[] call(InternalRow internalRow) {
+    String gcCatalogName = internalRow.getString(0);
+    String gcOutputBranchName = internalRow.getString(1);
+    String gcOutputTableIdentifier = internalRow.getString(2);
+    Map<String, String> nessieClientConfig = new HashMap<>();
+    MapData map = internalRow.getMap(3);
+    for (int i = 0; i < map.numElements(); i++) {
+      nessieClientConfig.put(
+          map.keyArray().getUTF8String(i).toString(), map.valueArray().getUTF8String(i).toString());
+    }
+    String runId = !internalRow.isNullAt(4) ? internalRow.getString(4) : null;
+    boolean dryRun = !internalRow.isNullAt(5) && internalRow.getBoolean(5);
+
+    IdentifiedResultsRepo identifiedResultsRepo =
+        new IdentifiedResultsRepo(
+            spark(), gcCatalogName, gcOutputBranchName, gcOutputTableIdentifier);
+
+    runId = updateRunId(gcOutputTableIdentifier, runId, identifiedResultsRepo);
+    FileIO fileIO = getFileIO(nessieClientConfig);
+    Set<String> expiredManifests = getExpiredManifests(runId, identifiedResultsRepo, fileIO);
+
+    Dataset<Row> expiredContents =
+        getExpiredContents(runId, identifiedResultsRepo, expiredManifests, fileIO);
+
+    expiredContents.persist();
+
+    if (!dryRun) {
+      expiredContents.foreach(
+          row -> {
+            List<String> files = row.getList(2);
+            files.forEach(fileIO::deleteFile);
```

Comment:
How does the whole thing behave, if one or more random files cannot be deleted?

Reply:
I will catch IO or Runtime exceptions and log the warning. I don't want to halt the whole expiry process.